catwalk/models/gpt.py

```diff
@@ -1,34 +1,161 @@
+            # To do this we need to find the index of the first/last token in the context.
+            # For example,
+            # Tokens:         PADDING CONTEXT QUESTION
+            # Token Type IDs:    0        0       1 (Used to find the last token of the context)
+            # Attention Mask:    0        1       1 (Used to find the first token of the context)
+
+            def get_token_bounds_for_context(sample_idx: int) -> Tuple[int, int]:
+                attention_mask = encodings["attention_mask"][sample_idx]
+                sequence_ids = token_type_map[sample_idx]
+
+                start_index = 0
+                while (attention_mask[start_index] == 0):
+                    start_index += 1
+
+                end_index = len(sequence_ids) - 1
+                while sequence_ids[end_index] != 0:
+                    end_index -= 1
+
+                return start_index, end_index
+
+            def contains_answer(sample_idx: int, answer_indices: Tuple[int, int]) -> bool:
+                start_index, end_index = get_token_bounds_for_context(
+                    sample_idx)
+                offset = offset_mapping[sample_idx]
+                return offset[start_index][0] <= answer_indices[0] and answer_indices[1] <= offset[end_index][1]
+
+            # Keep only the first sample for each instance that contains the answer.
+            for example_idx in sample_to_example_idx.keys():
+                answers = batch[example_idx].answers
+                for sample_idx in sample_to_example_idx[example_idx]:
+                    for answer_idx in range(len(answers["text"])):
+                        start_char = answers["answer_start"][answer_idx]
+                        end_char = start_char + \
+                            len(answers["text"][answer_idx].strip())
+                        if contains_answer(sample_idx, (start_char, end_char)):
+                            filtered_encodings["input_ids"].append(
+                                encodings["input_ids"][sample_idx])
+                            filtered_encodings["attention_mask"].append(
+                                encodings["attention_mask"][sample_idx])
+                            break # No need to check other answers for this example
+
+                    # check if we found a sample that contains the answer and break out of the loop
+                    if len(filtered_encodings["input_ids"])-1 == example_idx:
+                        break
+
+            with torch.inference_mode():
+                outputs = model.generate(input_ids=torch.stack(filtered_encodings["input_ids"]).to(device),
+                                         attention_mask=torch.stack(
+                                             filtered_encodings["attention_mask"]).to(device),
+                                         max_new_tokens=20,
+                                         pad_token_id=tokenizer.pad_token_id,
```

Comment:
Only a possibility but we could also have it terminate generation after some token. In the EAI tasks the task itself defines what that token is, but here I guess we'd have to make it a model arg. It's unfortunate those would be different, so I'm not committed to this but I just wanted to paste it here for reference. Feel free to reject the idea.

Also I think integrating this properly would also require purging padding tokens from the generated outputs because of how the batched generation handles the early termination.

```suggestion
                                         pad_token_id=tokenizer.pad_token_id,
                                         eos_token_id=tokenizer.encode('\n')[0],
```